packages/loader/container-loader/src/connectionHandler.ts

```diff
@@ -0,0 +1,889 @@
+     * about current or last connection (if there is no connection at the moment)
+    */
+     public get connectionProps(): ITelemetryProperties {
+        if (this.connection !== undefined) {
+            return {
+                connectionMode: this.connectionMode,
+                relayServiceAgent: this.connection.relayServiceAgent,
+                socketDocumentId: this.connection?.claims.documentId,
+            };
+        } else {
+            return {
+                // Report how many ops this client sent in last disconnected session
+                sentOps: this.clientSequenceNumber,
+            };
+        }
+    }
+
+    public shouldJoinWrite(): boolean {
+        // We don't have to wait for ack for topmost NoOps. So subtract those.
+        return this.clientSequenceNumberObserved < (this.clientSequenceNumber - this.trailingNoopCount);
+    }
+
+    /**
+     * Tells if container is in read-only mode.
+     * Data stores should listen for "readonly" notifications and disallow user
+     * making changes to data stores.
+     * Readonly state can be because of no storage write permission,
+     * or due to host forcing readonly mode for container.
+     * It is undefined if we have not yet established websocket connection
+     * and do not know if user has write access to a file.
+     */
+    private get readonly() {
+        if (this._forceReadonly) {
+            return true;
+        }
+        return this._readonlyPermissions;
+    }
+
+    public get readOnlyInfo(): ReadOnlyInfo {
+        const storageOnly = this.connection !== undefined && this.connection instanceof NoDeltaStream;
+        if (storageOnly || this._forceReadonly || this._readonlyPermissions === true) {
+            return {
+                readonly: true,
+                forced: this._forceReadonly,
+                permissions: this._readonlyPermissions,
+                storageOnly,
+            };
+        }
+
+        return { readonly: this._readonlyPermissions };
+    }
+
+    private static detailsFromConnection(connection: IDocumentDeltaConnection): IConnectionDetails {
+        return {
+            claims: connection.claims,
+            clientId: connection.clientId,
+            existing: connection.existing,
+            checkpointSequenceNumber: connection.checkpointSequenceNumber,
+            get initialClients() { return connection.initialClients; },
+            maxMessageSize: connection.serviceConfiguration.maxMessageSize,
+            mode: connection.mode,
+            serviceConfiguration: connection.serviceConfiguration,
+            version: connection.version,
+        };
+    }
+
+    constructor(
+        private readonly serviceProvider: () => IDocumentService | undefined,
+        private client: IClient,
+        reconnectAllowed: boolean,
+        private readonly logger: ITelemetryLogger,
+        private readonly props: IConnectionManagereFactoryArgs,
```

Comment:
Passing in as a single props is prob ok, but maybe consider splitting props out into individual private members rather than storing the props directly for legibility.